Every addUserToGroup has a removeUserFromGroup [GAWB-3099]

aednichols opened the pull request:
~~Kinda just want to see if tests pass with this in place.~~

This PR fixes one potential source of users accumulating unreasonable numbers of groups over multiple test runs (Rawls precondition failed/412).

Related, but not necessarily merging at the same time: https://github.com/broadinstitute/rawls/pull/844

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

aednichols committed:
Every addUserToGroup has a removeUserFromGroup

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/security/AuthDomainSpec.scala:
```diff
@@ -598,6 +598,8 @@ class AuthDomainSpec extends FreeSpec /*with ParallelTestExecution*/ with Matche
             api.billing.removeUserFromBillingProject(projectName, user.email, BillingProjectRole.Owner)
             checkNoAccess(user, projectName, workspaceName)
+
+            api.groups.removeUserFromGroup(groupName, user.email, GroupRole.Member)
```
These should be:
```
register cleanUp api.groups.removeUserFromGroup(groupName, user.email, GroupRole.Member)
```
and be done immediately after the call to `addUserToGroup`. You may also need to make use of `withCleanUp` (inside the `withGroup`) to make sure `removeUserFromGroup` happens before the group is deleted. I'm not sure how Google handles removing users from groups that have been deleted.

I'd be surprised if deleting the group doesn't also effectively wipe it from the list of groups a user belongs to. In fact, I felt a little silly removing members before deleting the group in `withGroup` but figured it didn't hurt.

Sorry for missing this clean-up.

breilly2 reviewed (approved):
👍  after minor change.

aednichols committed:
Brian's suggestions

In particular, if addUserToGroup throws an exception I think removeUserFromGroup might never be reached unless we use register cleanUp

breilly2 commented on automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/security/AuthDomainSpec.scala:
```diff
@@ -725,15 +734,19 @@ class AuthDomainSpec extends FreeSpec /*with ParallelTestExecution*/ with Matche
 
-            api.billing.removeUserFromBillingProject(projectName, user.email, BillingProjectRole.Owner)
-            checkNoAccess(user, projectName, workspaceName)
+              api.billing.addUserToBillingProject(projectName, user.email, BillingProjectRole.Owner)
```
A little unrelated but this maybe should also have:
```
register cleanUp api.billing.removeUserFromBillingProject(projectName, user.email, BillingProjectRole.Owner)
```

aednichols replied:
Replacing or in addition to
```
api.billing.removeUserFromBillingProject(projectName, user.email, BillingProjectRole.Owner)
```
on line 747?

aednichols replied:
Assuming replacing based on the pattern for the other one

breilly2 replied:
Actually it's in addition to because of the `checkNoAccess` at the end. I would expect that assertion to fail in the test's current form.

aednichols replied:
Gotcha, fell into mindless pattern recognition

aednichols commented:
We do get a stack trace on some of these when the group doesn't exist

```
WARN org.broadinstitute.dsde.firecloud.test.security.AuthDomainSpec - Error in clean-up function

broad-dsde-qa/AuthDomainSpec_create_598aa8ce-32ef-4884-a36d-d8e65471f61a does not exist
```

https://fc-jenkins.dsp-techops.broadinstitute.org/view/Test/job/fiab-test/2103/consoleFull

breilly2 commented:
That particular one is because the test deleted the workspace so there isn't anything to clean up. However it makes sense to have the clean up in case deleting the workspace from the test fails (for reasons other than the orch call is broken, such as broken UI).

aednichols commented:
jenkins retest

aednichols committed:
Register cleanup for billing project removal

rushtong reviewed (approved):
👍

MatthewBemis reviewed (approved):
:+1:

aednichols committed:
Restore operation required for test to pass

davidangb reviewed (approved):
:+1: code changes look good; I did not audit to see if this is comprehensive.

aednichols committed:
Merge remote-tracking branch 'origin/develop' into selenium-fix-groups

aednichols commented:
jenkins retest
